AssertionError: Unexpected node type: <class 'ast.Subscript'> on version > 23.2.13

Hi when I use flake8, I'm getting this exception `AssertionError: Unexpected node type: <class 'ast.Subscript'>`. I notice the error on version `23.2.13` and `23.3.12`. When I revert back to `23.1.20` the error goes away.

### Stack Trace

<details>
<summary> Expand </summary>

```python
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/flake8/checker.py", line 621, in _run_checks
    return checker.run_checks()
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/flake8/checker.py", line 531, in run_checks
    self.run_ast_checks()
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/flake8/checker.py", line 435, in run_ast_checks
    for (line_number, offset, text, _) in runner:
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/bugbear.py", line 61, in run
    visitor.visit(self.tree)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/bugbear.py", line 336, in visit
    super().visit(node)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/ast.py", line 407, in visit
    return visitor(node)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/bugbear.py", line 417, in visit_Module
    self.generic_visit(node)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/ast.py", line 415, in generic_visit
    self.visit(item)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/bugbear.py", line 336, in visit
    super().visit(node)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/ast.py", line 407, in visit
    return visitor(node)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/bugbear.py", line 482, in visit_ClassDef
    self.generic_visit(node)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/ast.py", line 415, in generic_visit
    self.visit(item)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/bugbear.py", line 336, in visit
    super().visit(node)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/ast.py", line 407, in visit
    return visitor(node)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/bugbear.py", line 475, in visit_FunctionDef
    self.generic_visit(node)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/ast.py", line 415, in generic_visit
    self.visit(item)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/bugbear.py", line 336, in visit
    super().visit(node)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/ast.py", line 407, in visit
    return visitor(node)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/bugbear.py", line 487, in visit_Try
    self.generic_visit(node)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/ast.py", line 415, in generic_visit
    self.visit(item)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/bugbear.py", line 336, in visit
    super().visit(node)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/ast.py", line 407, in visit
    return visitor(node)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/bugbear.py", line 360, in visit_ExceptHandler
    names = [_to_name_str(e) for e in good_handlers]
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/bugbear.py", line 360, in <listcomp>
    names = [_to_name_str(e) for e in good_handlers]
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/bugbear.py", line 259, in _to_name_str
    return _to_name_str(node.value) + "." + node.attr
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/bugbear.py", line 259, in _to_name_str
    return _to_name_str(node.value) + "." + node.attr
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/bugbear.py", line 257, in _to_name_str
    assert isinstance(node, ast.Attribute), f"Unexpected node type: {type(node)}"
AssertionError: Unexpected node type: <class 'ast.Subscript'>
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.9.16/x64/bin/flake8", line 8, in <module>
    sys.exit(main())
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/flake8/main/cli.py", line 22, in main
    app.run(argv)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/flake8/main/application.py", line 336, in run
    self._run(argv)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/flake8/main/application.py", line 325, in _run
    self.run_checks()
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/flake8/main/application.py", line 229, in run_checks
    self.file_checker_manager.run()
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/flake8/checker.py", line 250, in run
    self.run_parallel()
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/flake8/checker.py", line 217, in run_parallel
    for ret in pool_map:
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/multiprocessing/pool.py", line 448, in <genexpr>
    return (item for chunk in result for item in chunk)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/multiprocessing/pool.py", line 870, in next
    raise value
AssertionError: Unexpected node type: <class 'ast.Subscript'>
```

</details>

--- a/README.rst
+++ b/README.rst
@@ -325,6 +325,11 @@ MIT
 Change Log
 ----------
 
+Unreleased
+~~~~~~~~~~
+
+* B030: Fix crash on certain unusual except handlers (e.g. ``except a[0].b:``)
+
 23.3.12
 ~~~~~~~~
 

--- a/bugbear.py
+++ b/bugbear.py
@@ -250,15 +250,18 @@ def _check_redundant_excepthandlers(names, node):
 def _to_name_str(node):
     # Turn Name and Attribute nodes to strings, e.g "ValueError" or
     # "pkg.mod.error", handling any depth of attribute accesses.
+    # Return None for unrecognized nodes.
     if isinstance(node, ast.Name):
         return node.id
     if isinstance(node, ast.Call):
         return _to_name_str(node.func)
-    assert isinstance(node, ast.Attribute), f"Unexpected node type: {type(node)}"
-    try:
-        return _to_name_str(node.value) + "." + node.attr
-    except AttributeError:
-        return _to_name_str(node.value)
+    elif isinstance(node, ast.Attribute):
+        inner = _to_name_str(node.value)
+        if inner is None:
+            return None
+        return f"{inner}.{node.attr}"
+    else:
+        return None
 
 
 def names_from_assignments(assign_target):
@@ -345,19 +348,22 @@ class BugBearVisitor(ast.NodeVisitor):
             self.generic_visit(node)
             return
         handlers = _flatten_excepthandler(node.type)
-        good_handlers = []
+        names = []
         bad_handlers = []
         ignored_handlers = []
         for handler in handlers:
             if isinstance(handler, (ast.Name, ast.Attribute)):
-                good_handlers.append(handler)
+                name = _to_name_str(handler)
+                if name is None:
+                    ignored_handlers.append(handler)
+                else:
+                    names.append(name)
             elif isinstance(handler, (ast.Call, ast.Starred)):
                 ignored_handlers.append(handler)
             else:
                 bad_handlers.append(handler)
         if bad_handlers:
             self.errors.append(B030(node.lineno, node.col_offset))
-        names = [_to_name_str(e) for e in good_handlers]
         if len(names) == 0 and not bad_handlers and not ignored_handlers:
             self.errors.append(B029(node.lineno, node.col_offset))
         elif (

--- a/tests/b030.py
+++ b/tests/b030.py
@@ -32,3 +32,9 @@ try:
     pass
 except what_to_catch():  # ok
     pass
+
+
+try:
+    pass
+except a.b[1].c:  # ok
+    pass

